biohansel/create/cluster_generator.py

```diff
@@ -0,0 +1,236 @@
+from typing import Dict, Set, List
+
+import numpy as np
+import pandas as pd
+import scipy as sp
+
+
+from collections import defaultdict
+from scipy.cluster.hierarchy import fcluster, linkage
+from scipy.spatial.distance import pdist
+
+
+def find_clusters(df: pd.DataFrame, min_group_size: int) -> Dict[str, int]:
+    """
+    Takes in a vcf file and creates clusters from the scipy hierarchy clustering algorithm
+    
+    Example:
+    '/path/example.vcf' -> {'mysnpsSRR2323':'1', 'mysnpsSRR232323':'2'}
+    
+    Args:
+        df: contains the vcf information in the DataFrame format
+        min_group_size: the minimum child group size for each new subtype branching point from the parent group
+
+    Returns:
+        cluster_dict: a dictionary indicating the cluster membership of each of the supplied genomes in
+                the vcf file
+    """
+
+    distance_matrix = compute_distance_matrix(df)
+    clustering_array = create_linkage_array(distance_matrix)
+
+    flat_clusters = output_flat_clusters(clustering_array, df.columns, distance_matrix, min_group_size)
+
+    return flat_clusters
+
+
+def compute_distance_matrix(df: pd.DataFrame) -> np.ndarray:
+    """Takes in a binary SNV state DataFrame and outputs a distance matrix using the hamming method
+
+    Args:
+        df: the DataFrame that contains only the samples' binary data
+
+    Returns:
+        a matrix of pair-wise distances between samples
+    """
+
+    return sp.spatial.distance.pdist(
+        df.transpose(), metric='hamming')
+
+
+def create_linkage_array(distance_matrix: np.ndarray) -> np.ndarray:
+    """Takes in a distance matrix and outputs a hierarchical clustering linkage array
+
+    Args:
+        distance_matrix: a matrix of pair-wise distances between samples
+
+    Returns:
+        clustering_array: a hierarchical clustering linkage array that is calculated from the distance matrix
+    """
+
+    clustering_array = linkage(distance_matrix, method='complete')
+
+    return clustering_array
+
+
+def output_flat_clusters(clustering_array: np.ndarray, genomes_only: List, distance_matrix: np.ndarray,
```

Comment:
This function isn't returning the flat clusters at various distance threshold, instead, it is returning the hierarchical clusters. Please split this function up into its component parts so that we can output the intermediate output to show how the final HC clusters are derived.

Reply:
yes this has been addressed in 94b2c0213878ff804e065eb970dc3023030acfcc